torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -3878,65 +4024,77 @@ def bwd(ctx, grad, x):
-        # tensors/attrs created in the fwd tracer.
-
-        if isinstance(fwd_out, variables.BaseListVariable):
-            bwd_args = [ctx, *fwd_out.items]
+        bwd_args = []
+        if isinstance(fwd_out, variables.TensorVariable):
+            bwd_args.append(fwd_out)
         else:
-            bwd_args = [ctx, fwd_out]
-
-        bwd_src = AttrSource(self.parent_source, member="backward")
-        if isinstance(self.bwd_graph, types.FunctionType):
-            bwd_fn = UserFunctionVariable(self.bwd_graph, source=bwd_src)
-        elif isinstance(self.bwd_graph, types.MethodType):
-            bwd_fn = UserMethodVariable(
-                self.bwd_graph.__func__,
-                UserDefinedClassVariable(self.bwd_graph.__class__),
-                source=bwd_src,
-            )
-            bwd_args = [bwd_fn.obj, *bwd_args]
-        else:
-            unimplemented(
-                gb_type="autograd.Function.apply: non-function or method backward",
-                context=str(self.bwd_graph),
-                explanation="Expected backward function to be a function or method.",
-                hints=[],
-            )
+            assert isinstance(fwd_out, variables.BaseListVariable)
+            for i in fwd_out.items:
+                if isinstance(i, variables.TensorVariable):
+                    bwd_args.append(i)
+                else:
+                    bwd_args.append(ConstantVariable.create(None))
+
+        bwd_fn, bwd_args = self.prepare_fn_vt(ctx, "backward", bwd_args)
 
         def is_strict_for(v: VariableTracker):
             if isinstance(v, variables.TensorVariable):
                 # we can be more lax for stuff from forward
                 return v.proxy.tracer is not fwd_tracer
             return True
 
+        # automatic_with_forced_inputs relies on the function arg names to
+        # create a new proxy. Also, it will always INSERT a tensor placeholder
+        # as input, even though it might not be used in the graph. This allows
+        # us to make a mapping for the backward graph.
         with (
             tx.output.subtracer(fwd_fn, fwd_tracer),
             tx.strict_translation_mode(is_strict_for),
         ):
             try:
-                (bwd_out, _), bwd_graph, bwd_freevars = speculate_subgraph(
-                    tx,
-                    bwd_fn,
-                    bwd_args,
-                    kwargs,
-                    "autograd.Function",
-                    enable_grad=False,
-                    set_subgraph_inputs="manual",
-                    restore_side_effects=False,
-                    tracer=bwd_tracer,
+                bwd_out, bwd_graph, bwd_freevars, bwd_graph_output_vts = (
+                    speculate_subgraph_with_auto_output_flattening(
+                        tx,
+                        bwd_fn,
+                        bwd_args,
+                        {},
+                        "autograd.Function",
+                        enable_grad=False,
```

Comment:
hmm, didn't realize the code did enable_grad=False on backward before, feel free to ignore previous comment.
I'm not completely sure this is correct but it is preexisting so should be fine